File: timigs-android/rust/src/models/task.rs
//! Task/Goal model

use chrono::{DateTime, Utc};
use flutter_rust_bridge::frb;
use serde::{Deserialize, Serialize};

/// Task status enum
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[frb(dart_metadata=("freezed"))]
pub enum TaskStatus {
    Active,
    Completed,
    Paused,
    Cancelled,
}

impl Default for TaskStatus {
    fn default() -> Self {
        Self::Active
    }
}

/// Usage goal/task for an app
#[derive(Debug, Clone, Serialize, Deserialize)]
#[frb(dart_metadata=("freezed"))]
pub struct Task {
    pub id: i64,
    pub app_name: String,
    pub description: Option<String>,
    pub goal_seconds: i64,
    pub created_at: DateTime<Utc>,
    pub status: TaskStatus,
    pub title_filter: Option<String>,
}

impl Task {
    pub fn is_active(&self) -> bool {
        self.status == TaskStatus::Active
    }

    /// Calculate progress percentage given actual seconds
    pub fn progress_percentage(&self, actual_seconds: i64) -> f64 {
        if self.goal_seconds == 0 {
            100.0
        } else {
            ((actual_seconds as f64 / self.goal_seconds as f64) * 100.0).min(100.0)
        }
    }
}

/// Input for creating a new task
#[derive(Debug, Clone, Serialize, Deserialize)]
#[frb(dart_metadata=("freezed"))]
pub struct NewTask {
    pub app_name: String,
    pub description: Option<String>,
    pub goal_seconds: i64,
    pub title_filter: Option<String>,
}

File: timigs-android/lib/models/task.dart
class Task {
  final int id;
  final String appName;
  final String? description;
  final int goalSeconds;
  final DateTime createdAt;
  final String status; // 'active', 'completed', 'paused'
  final String? titleFilter;

  Task({
    required this.id,
    required this.appName,
    this.description,
    required this.goalSeconds,
    required this.createdAt,
    required this.status,
    this.titleFilter,
  });

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'app_name': appName,
      'description': description,
      'goal_seconds': goalSeconds,
      'created_at': createdAt.toIso8601String(),
      'status': status,
      'title_filter': titleFilter,
    };
  }

  factory Task.fromMap(Map<String, dynamic> map) {
    return Task(
      id: map['id'] as int,
      appName: map['app_name'] as String,
      description: map['description'] as String?,
      goalSeconds: map['goal_seconds'] as int,
      createdAt: DateTime.parse(map['created_at'] as String),
      status: map['status'] as String,
      titleFilter: map['title_filter'] as String?,
    );
  }

  String get formattedGoal {
    final hours = goalSeconds ~/ 3600;
    final minutes = (goalSeconds % 3600) ~/ 60;
    
    if (hours > 0) {
      return '${hours}h ${minutes}m';
    } else {
      return '${minutes}m';
    }
  }
}
